### PR TITLE
RPC: Add child transactions to getrawmempool verbose output

### PR DESCRIPTION
**Pull request by conscott**
`bitcoin-cli getrawmempool true` only lists a transaction's parents in the `depends` field. This change adds a `spentby` field to the json response, which lists the transaction's children in the mempool.

Currently the only way to find child transactions is to use `getrawmempool` or make another call to `getmempooldescendants` and search the response for transactions that list the parent_txid in the `depends` list, which is inefficient. 

This change allows direct lookup of children. 

Example Output
```
  "9a9b5733c0d89f207908cfa3fe17809bee71f629aa095c9f8754524e29e98ba4": {
    ...other geterawmempool data...
    "wtxid": "9a9b5733c0d89f207908cfa3fe17809bee71f629aa095c9f8754524e29e98ba4",
    "depends": [
      "bdd92851d5766a42aeb62af667bb422a116cab4e032bba5e3dd6efe5b4b40aa0"
    ],
    "spentby": [
      "dc5d3ec388a9121421208738a041ac30a22163bc2e17758f2275b6c51a15ba7b"
    ]
  },
```

### CODE REVIEW

**Comment by esotericnonsense** [comment]
This is cool. I'd thought about adding a similar concept to getblocktemplate (or some sort of package number or something).

Have you given any thought to performance / done any benchmarking? I'm particularly thinking of what happens when the mempool is in the hundreds of megs as it was not too long ago.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -406,6 +409,24 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
     }
 
     info.pushKV("depends", depends);
+
+    std::set<std::string> setSpent;
+    for (unsigned int i = 0; i < tx.vout.size(); i++) {
+        auto it = mempool.mapNextTx.find(COutPoint(tx.GetHash(), i));
```
Comment (esotericnonsense) [inline src/rpc/blockchain.cpp]: Does this repeat tx.GetHash() for every vout?
Possibly makes sense to do it outside of the loop?
(Maybe the compiler is more intelligent than I'm imagining)

**Comment by dcousens** [comment]
NACK, atleast, not by default,  this would vastly increase the cost of this RPC call.

**Comment by conscott** [comment]
Per suggestion of @dcousens and @esotericnonsense I will perform some timing tests and pull this out into an optional parameter if it's hampering performance significantly.

**Comment by sipa** [comment]
I looks like it's a single map lookup per reported transaction output, which I don't expect to be all that expensive.

Numbers are better, of course.

**Comment by dcousens** [comment]
revoking my NACK,  I initially mis-interpreted the extent of the added data.
I'd instead suggest to additionally add the `vout` which spends the TX, not only the `txid`.

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (promag) [reply src/rpc/blockchain.cpp]: It's fine as is, see `CTransaction::GetHash()`.

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -372,6 +372,9 @@ std::string EntryDescriptionString()
            "    \"wtxid\" : hash,         (string) hash of serialized transaction, including witness data\n"
            "    \"depends\" : [           (array) unconfirmed transactions used as inputs for this transaction\n"
            "        \"transactionid\",    (string) parent transaction id\n"
+           "       ... ]\n"
+           "    \"spentby\" : [          (array) unconfirmed transactions spending outputs from this transaction\n"
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Looks like alignment is bad.

**Review by promag — COMMENTED** [review]
Concept ACK.

See `doc/developer-notes.md` regarding style for new code.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -406,6 +409,24 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
...
+            setSpent.insert(tx_hash.ToString());
+    }
+
+    UniValue spent(UniValue::VARR);
+    for (const std::string& dep : setSpent)
+    {
+        spent.push_back(dep);
+    }
+
+    info.pushKV("spentby", spent);
```
[the remark this answers is not in the document]
Comment (conscott) [reply src/rpc/blockchain.cpp]: Agreed. I have updated the tests to coverage all these cases.

**Comment by conscott** [comment]
Thanks for review @esotericnonsense @promag and @dcousens 

I have updated commits from your comments. 

It is correct that this just adds a map lookup per transactions, but I went ahead and tested performance anyway. This seems to average out to about 3% increase in total rpc call time testing with an 18k transaction mempool full of parent/child relations. 

The details are that I [hacked together a test]( https://github.com/bitcoin/bitcoin/commit/55a9289c82a1c088ae2be54452cfc1c4183172ff) to generate a mempool of 18k transactions with long chains of unconfirmed transactions that excercise child/parent relations for `getrawmempool` output. Once the mempool in generated there is a 5 minute sleep, so in a loop I could call `time bitcoin-cli -datadir=<test_dir> getrawmempool` with / without descendant pointers (verbose flag was reset to trigger this). I realize this is pretty kludgy, but just wanted to get a rough idea if something really bad was happening. 

@dcousens I am not sure we want to add the `vout` of Tx in the spentby list. A child transaction can spend multiple outputs of the same parent, so representing this will add a bit more complexity to the output. This also applies to the `depends` list, in that the current transaction could be spending multiple outputs for a single parent, and I think we want to keep the `spends` / `depends` format about the same. What are your thoughts?

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -406,6 +409,26 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
     }
 
     info.pushKV("depends", depends);
+
+    std::set<std::string> set_spent;
```
Comment (promag) [inline src/rpc/blockchain.cpp]: `std::set<uint256>`?

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -406,6 +409,26 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
...
+    for (unsigned int i = 0; i < tx.vout.size(); i++) {
+        auto it = mempool.mapNextTx.find(COutPoint(tx.GetHash(), i));
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Not sure if it matters, but how about `const auto& it = `?

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -406,6 +409,26 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
...
+        if (it == mempool.mapNextTx.end()) {
+            continue;
+        }
+        const uint256& tx_hash = it->second->GetHash();
+        if (mempool.exists(tx_hash)) {
+            set_spent.insert(tx_hash.ToString());
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Avoid loop below by taking advantage of `insert()` return value:
```cpp
if (set_spent.insert(tx_hash).second) {
    spent.push_back(tx_hash.ToString());
}
```

**Review by jamesob — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/12479/commits/8ad782d7071b8e1042e27a9c147e935b457c46b6

File: src/rpc/blockchain.cpp [hunk 7]
```diff
@@ -406,6 +409,23 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
...
+
+    std::set<uint256> set_spent;
+    UniValue spent(UniValue::VARR);
+    for (unsigned int i = 0; i < tx.vout.size(); i++) {
+        const auto& it = mempool.mapNextTx.find(COutPoint(tx.GetHash(), i));
+        if (it == mempool.mapNextTx.end()) {
+            continue;
+        }
+        const uint256& tx_hash = it->second->GetHash();
+        if (mempool.exists(tx_hash)) {
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Is this test really necessary? If `mapNextTx` contains the outpoint, then the spending transaction should be in the mempool? See `CTxMemPool::addUnchecked()`.

Therefore, maybe:
 - replace with `assert(mempool.exists(tx_hash))`
 - or test and throw an error if not in the mempool (or log something)
 - or even remove the check.

IMO not being in `mempool.mapTx` means that something is broken so I would choose the assert.

cc @morcos @sdaftuar

File: src/rpc/blockchain.cpp [hunk 8]
```diff
@@ -406,6 +409,23 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
     }
 
     info.pushKV("depends", depends);
+
+    std::set<uint256> set_spent;
+    UniValue spent(UniValue::VARR);
+    for (unsigned int i = 0; i < tx.vout.size(); i++) {
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Nit, `++i`.

**Review by promag — COMMENTED** [review]
utACK 8ad782d. I would like to get feedback from my comment below though.

BTW 2nd commit message is very long.

File: src/rpc/blockchain.cpp (hunk 8 above)
Comment (junfx) [inline src/rpc/blockchain.cpp]: This for loop iterates tx.vout, does it equal to:
`auto entryit = mempool.mapTx.find(tx.GetHash());`
`auto children = mempool.GetMemPoolChildren(entryit);`
?

File: src/rpc/blockchain.cpp (hunk 7 above)
Comment (conscott) [reply src/rpc/blockchain.cpp]: I agree. This was a paranoia check and I debated whether this should be an assert. I think I will update as such.

File: src/rpc/blockchain.cpp (hunk 8 above)
Comment (promag) [reply src/rpc/blockchain.cpp]: The index is needed below, to construct the outpoint from the output.

Comment (promag) [reply src/rpc/blockchain.cpp]: Right, I've overlooked @junfx suggestion.

**Comment by morcos** [comment]
ACK 1dfb4e7

**Comment by dcousens** [comment]
> This also applies to the depends list, in that the current transaction could be spending multiple outputs for a single parent, and I think we want to keep the spends / depends format about the same. What are your thoughts?

@conscott I think that is fine,  I guess it depends on the verbosity required for your application.  The few places I can see this being useful,  I would appreciate the `vout` information (as is, I simply download everything).

**Comment by sipa** [comment]
utACK 1dfb4e7d753e9282c89d55bde358b8ad96d3bfc2

**Pull request merged**